### DISCUSSION TITLE
Are pre_build and post_build rule changes persistent?

### DISCUSSION THREAD

**Discussion by IcedQuinn** [thought-machine/please#2101] [Q&A] [answered]
For example in Ninja it is possible for a build rule to output a list of dependencies the file has (such as what headers a C file used.) It then load that list of dependencies and stores them in the cache for future use. This allows it to suss out which C files need to be recompiled whenever some headers are changed without each one being explicitly specified.

I noticed there are `pre_build` and `post_build` rules which are allowed to modify a build rule. Are those changes cached anywhere so rules are able to do similar dependency detection and incremental rebuilding?

**Comment by IcedQuinn** [comment]
Looking in to this I checked if C/C++ targets are rebuilt by looking at the dependency graph for each .c file.

- They are not.
- Changes to a particular .cc file results in just that .cc file being recompiled
- However changing a header triggers every .cc to be rebuilt even the ones which did not depend on it

**Comment by IcedQuinn** [comment, marked as answer]
I don't see anywhere where the build rules would be cached so it looks like no, they are not. It looks like you would have to use this out of band and put the dependencies in the BUILD file.

**Comment by peterebden** [comment]
Correct, they are not persisted - the function is re-evaluated on each plz invocation. However the stdout of an action that's passed to the post-build function _is_ persisted in the cache, so the actual commands don't need to get rerun, and hence they don't prevent incremental rebuilds.
